stacks-signer/src/crypto/frost.rs

```diff
@@ -0,0 +1,780 @@
+                    .insert(dkg_public_share.party_id, dkg_public_share.clone());
+
+                info!(
+                    "DKG round #{} DkgPublicShare from party #{}",
+                    dkg_public_share.dkg_id, dkg_public_share.party_id
+                );
+            }
+            _ => {}
+        }
+        if self.ids_to_await.is_empty() {
+            // Calculate the aggregate public key
+            let key = self
+                .dkg_public_shares
+                .iter()
+                .fold(Point::default(), |s, (_, dps)| s + dps.public_share.A[0]);
+            // check to see if aggregate public key has even y
+            if key.has_even_y() {
+                info!("Aggregate public key has even y coord!");
+                info!("Aggregate public key: {}", key);
+                self.aggregate_public_key = key;
+                self.move_to(State::DkgPrivateDistribute)?;
+            } else {
+                warn!("DKG Round #{} Failed: Aggregate public key does not have even y coord, re-running dkg.", self.current_dkg_id);
+                self.move_to(State::DkgPublicDistribute)?;
+            }
+            self.ids_to_await = (0..self.total_signers).collect();
+        }
+        Ok(())
+    }
+
+    fn gather_dkg_end(&mut self, message: &Message) -> Result<(), Error> {
+        info!(
+            "DKG Round #{}: waiting for Dkg End from signers {:?}",
+            self.current_dkg_id, self.ids_to_await
+        );
+        if let MessageTypes::DkgEnd(dkg_end_msg) = &message.msg {
+            self.ids_to_await.remove(&dkg_end_msg.signer_id);
+            info!(
+                "DKG_End round #{} from signer #{}. Waiting on {:?}",
+                dkg_end_msg.dkg_id, dkg_end_msg.signer_id, self.ids_to_await
+            );
+        }
+
+        if self.ids_to_await.is_empty() {
+            self.ids_to_await = (0..self.total_signers).collect();
+            self.move_to(State::Idle)?;
+        }
+        Ok(())
+    }
+
+    fn request_nonces(&mut self) -> Result<Message, Error> {
+        info!(
+            "Sign Round #{} Nonce round #{} Requesting Nonces",
+            self.current_sign_id, self.current_sign_nonce_id, 
+        );
+        let nonce_request = NonceRequest {
+            dkg_id: self.current_dkg_id,
+            sign_id: self.current_sign_id,
+            sign_nonce_id: self.current_sign_nonce_id,
+        };
+        let nonce_request_msg = Message {
+            sig: nonce_request.sign(&self.message_private_key).expect(""),
+            msg: MessageTypes::NonceRequest(nonce_request),
+        };
+        self.ids_to_await = (0..self.total_signers).collect();
+        self.move_to(State::NonceGather)?;
+        Ok(nonce_request_msg)
+    }
+
+    fn gather_nonces(&mut self, message: &Message) -> Result<(), Error> {
+        match &message.msg {
+            MessageTypes::NonceResponse(nonce_response) => {
+                // TODO: check sign_id and sign_nonce_id
```

Comment:
Does this need to be addressed in order for the code to be robust against a malicious signer?

Reply:
It needs to be addressed for general robustness, and specifically we will need it for the meta-protocols FIRE/ROAST (where we handle byzantine signers).  It's fine as is for now, but if you want I can do an initial robustness pass before merging this PR.

Reply:
I went ahead and implemented this, which should make it less work for when I implement `FIRE`.

@jferrant we got another TODO: incrementing `dkg_public_id` if public distribution fails (due to odd y coord on agg public key) doesn't work because we don't have a `DkgPublicBegin` message; we just re-send `DkgBegin`, so we're always sending `dkg_public_id: 0` from the `SigningRound`.

Reply:
Oh missed this. Adding https://github.com/stacks-network/stacks-blockchain/issues/3924